S3RepositoryThirdPartyTests testCleanup failure in CI

In CI, the S3RepositoryThirdPartyTests testCleanup test failed. I checked to ensure that this is not the same issue as #52191, where multiple tests ran concurrently and it does not appear to be the case in this instance.

Reproduce line:

```
./gradlew ':plugins:repository-s3:thirdPartyTest' --tests "org.elasticsearch.repositories.s3.S3RepositoryThirdPartyTests.testCleanup" \
  -Dtests.seed=A30B885AB17443F3 \
  -Dtests.security.manager=true \
  -Dtests.locale=sq \
  -Dtests.timezone=America/Cayman \
  -Dcompiler.java=13
```

```
ElasticsearchTimeoutException[java.util.concurrent.TimeoutException: Timeout waiting for task.]; nested: TimeoutException[Timeout waiting for task.];
	at org.elasticsearch.common.util.concurrent.FutureUtils.get(FutureUtils.java:78)
	at org.elasticsearch.action.support.AdapterActionFuture.actionGet(AdapterActionFuture.java:61)
	at org.elasticsearch.action.support.AdapterActionFuture.actionGet(AdapterActionFuture.java:55)
	at org.elasticsearch.repositories.blobstore.BlobStoreTestUtil.assertConsistency(BlobStoreTestUtil.java:140)
	at org.elasticsearch.repositories.AbstractThirdPartyRepositoryTestCase.assertConsistentRepository(AbstractThirdPartyRepositoryTestCase.java:260)
	at org.elasticsearch.repositories.s3.S3RepositoryThirdPartyTests.lambda$assertConsistentRepository$1(S3RepositoryThirdPartyTests.java:94)
	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:893)
	at org.elasticsearch.repositories.s3.S3RepositoryThirdPartyTests.assertConsistentRepository(S3RepositoryThirdPartyTests.java:94)
	at org.elasticsearch.repositories.AbstractThirdPartyRepositoryTestCase.testCleanup(AbstractThirdPartyRepositoryTestCase.java:218)
```

```
com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=37, name=elasticsearch[node_s_0][generic][T#4], state=RUNNABLE, group=TGRP-S3RepositoryThirdPartyTests]
Caused by: java.lang.AssertionError: expected:<0> but was:<1>
	at __randomizedtesting.SeedInfo.seed([A30B885AB17443F3]:0)
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at org.elasticsearch.repositories.blobstore.BlobStoreTestUtil.assertIndexGenerations(BlobStoreTestUtil.java:147)
	at org.elasticsearch.repositories.blobstore.BlobStoreTestUtil.lambda$assertConsistency$0(BlobStoreTestUtil.java:129)
```

--- a/test/framework/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreTestUtil.java
+++ b/test/framework/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreTestUtil.java
@@ -112,27 +112,35 @@ public final class BlobStoreTestUtil {
      *                 of this assertion must pass an executor on those when using such an implementation.
      */
     public static void assertConsistency(BlobStoreRepository repository, Executor executor) {
-        final PlainActionFuture<Void> listener = PlainActionFuture.newFuture();
-        executor.execute(ActionRunnable.run(listener, () -> {
-            final BlobContainer blobContainer = repository.blobContainer();
-            final long latestGen;
-            try (DataInputStream inputStream = new DataInputStream(blobContainer.readBlob("index.latest"))) {
-                latestGen = inputStream.readLong();
-            } catch (NoSuchFileException e) {
-                throw new AssertionError("Could not find index.latest blob for repo [" + repository + "]");
+        final PlainActionFuture<AssertionError> listener = PlainActionFuture.newFuture();
+        executor.execute(ActionRunnable.supply(listener, () -> {
+            try {
+                final BlobContainer blobContainer = repository.blobContainer();
+                final long latestGen;
+                try (DataInputStream inputStream = new DataInputStream(blobContainer.readBlob("index.latest"))) {
+                    latestGen = inputStream.readLong();
+                } catch (NoSuchFileException e) {
+                    throw new AssertionError("Could not find index.latest blob for repo [" + repository + "]");
+                }
+                assertIndexGenerations(blobContainer, latestGen);
+                final RepositoryData repositoryData;
+                try (InputStream blob = blobContainer.readBlob("index-" + latestGen);
+                     XContentParser parser = XContentType.JSON.xContent().createParser(NamedXContentRegistry.EMPTY,
+                         LoggingDeprecationHandler.INSTANCE, blob)) {
+                    repositoryData = RepositoryData.snapshotsFromXContent(parser, latestGen);
+                }
+                assertIndexUUIDs(blobContainer, repositoryData);
+                assertSnapshotUUIDs(repository, repositoryData);
+                assertShardIndexGenerations(blobContainer, repositoryData.shardGenerations());
+                return null;
+            } catch (AssertionError e) {
+                return e;
             }
-            assertIndexGenerations(blobContainer, latestGen);
-            final RepositoryData repositoryData;
-            try (InputStream blob = blobContainer.readBlob("index-" + latestGen);
-                 XContentParser parser = XContentType.JSON.xContent().createParser(NamedXContentRegistry.EMPTY,
-                     LoggingDeprecationHandler.INSTANCE, blob)) {
-                repositoryData = RepositoryData.snapshotsFromXContent(parser, latestGen);
-            }
-            assertIndexUUIDs(blobContainer, repositoryData);
-            assertSnapshotUUIDs(repository, repositoryData);
-            assertShardIndexGenerations(blobContainer, repositoryData.shardGenerations());
         }));
-        listener.actionGet(TimeValue.timeValueMinutes(1L));
+        final AssertionError err = listener.actionGet(TimeValue.timeValueMinutes(1L));
+        if (err != null) {
+            throw new AssertionError(err);
+        }
     }
 
     private static void assertIndexGenerations(BlobContainer repoRoot, long latestGen) throws IOException {

